README.md
# Metricsd

Metricsd is a pure Ruby client library for the [MetricsD](https://github.com/kpumuk/metricsd) server.

## Installation

Add the "metricsd" gem to your `Gemfile`.

    gem 'metricsd'

And run `bundle install` command.

## Getting started

You can configure Metricsd connection parameters by accessing attributes of `Metricsd` module:

    Metricsd.server_host = 'metrics.local'
    Metricsd.server_port = 6311

There are few more setting, please check project documentation.

Now you should be able to record you metrics:

    # Record success hit
    Metricsd::Client.record_success("api.docs.upload")
    # Record failure hit
    Metricsd::Client.record_failure("api.docs.upload")
    # Record timing info
    Metricsd::Client.record_time("api.docs.upload", 0.14)
    # Record complete success hit info (count + timing)
    Metricsd::Client.record_hit("api.docs.upload", true, 0.14)
    # Record an integer value
    Metricsd::Client.record_value("user.password.size", 15)
    Metricsd::Client.record_value("user.age", 26)

You can combine you metrics to send them in a single network packet for performance reason:

    # Send all database pool stats
    Metricsd::Client.record_values({
      'db.pool.reserved'  => db_stats[:reserved],
      'db.pool.available' => db_stats[:available],
      'db.pool.pending'   => db_stats[:pending],
    }, :group => 'doc_timestamp')

You can specify message source using :source => 'src' option. In this case you will be able to see summary graphs and graphs per source:

    # Generate graphs for all tables, and each single table.
    Metricsd::Client.record_success("hbase.reads", :source => @hbase_table)

By default only summary statistics is calculated. You can enable per-host graphs by specifying the appropriate source:

    # Generate summary graph for all hosts, and graphs for each single host.
    Metricsd::Client.record_success("hbase.reads", :source => Metricsd::Client.source)
    # ... or you can pass an empty string with the same effect.
    Metricsd::Client.record_success("hbase.reads", :source => '')

You can group your metrics using :group option. In this case metrics will be displayed together on the summary page.

    # Group metrics using :group option.
    Metricsd::Client.record_success("reads", :source => @hbase_table, :group => 'hbase')
    # Group metrics using special syntax "group$metric".
    Metricsd::Client.record_success("hbase$reads", :source => @hbase_table)

## More info

Check the [Project Documentation](http://rubydoc.info/gems/metricsd/) or check the tests to find out how to use this client.


CHANGELOG.md
## 0.2.6 (September 20, 2012)

Features:

  - Added Client#record_status method

## 0.2.5 (November 16, 2011)

Bugfixes:

  - Do not fail when MetricsD host could not be resolved

## 0.2.4 (August 11, 2011)

Features:

  - Time metrics should be recorded with ".time" suffix (was "\_time"), status metrics - with ".status" suffix (was "_count")
  - Added Client#record methods (alias of record_value)
  - Changed group separator to "." (was "$")
  - Removed separator option (it was a temporary solution)

## 0.2.3 (July 27, 2011)

Features:

  - Added ability to specify default group

## 0.2.2 (July 27, 2011)

Bugfixes:

  - Fixed problem in connection establishing code

## 0.2.1 (July 27, 2011)

Bugfixes:

  - Fixed problem with safe_send under Goliath

## 0.2.0 (July 27, 2011)

Features:

  - Added an option to enable or disable client

## 0.1.0 (July 26, 2011)

Features:

  - Complete MetricsD protocol implementation
  - Allows to send multiple metrics in a single network packet
  - Splits metrics to several packets of 250 bytes or less
  - Allows to specify metrics parts separator (temporary solution until all the metrics will be migrated to "." separator)
  - Logs network problems to logger and never throws errors like this.
  - 100% test coverage


spec/client_spec.rb
require 'bundler/setup'
require 'metricsd'

describe Metricsd::Client do
  before :each do
    @socket = mock('Socket')
    Metricsd::Client.stub!(:collector_socket => @socket)
    # Global settings
    Metricsd.reset_defaults!
    Metricsd.source = 'test'
  end

  describe 'connection' do
    it 'should not throw, but log exceptions on failure' do
      @socket.should_receive(:send).and_raise(Errno::ECONNREFUSED.new('exception from test'))
      Metricsd.logger.should_receive(:error).once.with(match(/exception from test/))
      Metricsd::Client.record_value('custom.metric', 5)
    end

    it 'should not send anything to the socket if disabled' do
      Metricsd.disable!
      @socket.should_not_receive(:send)
      Metricsd::Client.record_value('custom.metric', 5)
    end

    it 'should create and cache UDPSocket in collector_socket method' do
      Metricsd::Client.unstub!(:collector_socket)
      sock = Metricsd::Client.send(:collector_socket)
      sock.should be_a(UDPSocket)
      Metricsd::Client.send(:collector_socket).should be(sock)
    end
  end

  describe '.record_hit' do
    it 'should send two metrics in a single packet' do
      @socket.should_receive(:send).with('all@custom.metric.status:1;all@custom.metric.time:450', 0)
      Metricsd::Client.record_hit('custom.metric', true, 0.45)
    end

    it 'should handle is_success=false' do
      @socket.should_receive(:send).with('all@custom.metric.status:-1;all@custom.metric.time:450', 0)
      Metricsd::Client.record_hit('custom.metric', false, 0.45)
    end

    it 'should apply group to both the metrics' do
      @socket.should_receive(:send).with('all@test.custom.metric.status:1;all@test.custom.metric.time:450', 0)
      Metricsd::Client.record_hit('custom.metric', true, 0.45, :group => 'test')
    end

    it 'should apply source if empty string passed' do
      @socket.should_receive(:send).with('test@custom.metric.status:1;test@custom.metric.time:450', 0)
      Metricsd::Client.record_hit('custom.metric', true, 0.45, :source => '')
    end

    it 'should apply source if specified' do
      @socket.should_receive(:send).with('test2@custom.metric.status:1;test2@custom.metric.time:450', 0)
      Metricsd::Client.record_hit('custom.metric', true, 0.45, :source => 'test2')
    end
  end

  describe '.record_success' do
    it 'should record successes' do
      @socket.should_receive(:send).with('all@custom.metric.status:1', 0)
      Metricsd::Client.record_success('custom.metric')
    end

    it 'should apply group if specified' do
      @socket.should_receive(:send).with('all@test.custom.metric.status:1', 0)
      Metricsd::Client.record_success('custom.metric', :group => 'test')
    end

    it 'should apply source if empty string passed' do
      @socket.should_receive(:send).with('test@custom.metric.status:1', 0)
      Metricsd::Client.record_success('custom.metric', :source => '')
    end

    it 'should apply source specified' do
      @socket.should_receive(:send).with('test2@custom.metric.status:1', 0)
      Metricsd::Client.record_success('custom.metric', :source => 'test2')
    end
  end

  describe '.record_failure' do
    it 'should record failures' do
      @socket.should_receive(:send).with('all@custom.metric.status:-1', 0)
      Metricsd::Client.record_failure('custom.metric')
    end

    it 'should apply group if specified' do
      @socket.should_receive(:send).with('all@test.custom.metric.status:-1', 0)
      Metricsd::Client.record_failure('custom.metric', :group => 'test')
    end

    it 'should apply source if empty string passed' do
      @socket.should_receive(:send).with('test@custom.metric.status:-1', 0)
      Metricsd::Client.record_failure('custom.metric', :source => '')
    end

    it 'should apply source specified' do
      @socket.should_receive(:send).with('test2@custom.metric.status:-1', 0)
      Metricsd::Client.record_failure('custom.metric', :source => 'test2')
    end
  end

  describe '.record_time' do
    it 'should record time if specified' do
      @socket.should_receive(:send).with('all@custom.metric.time:450', 0)
      Metricsd::Client.record_time('custom.metric', 0.45)
    end

    it 'should yield a block if time is not specified' do
      @socket.should_receive(:send).with(match(/all@custom.metric.time:1\d{2}/), 0)
      yielded = false
      Metricsd::Client.record_time('custom.metric') do
        yielded = true
        sleep 0.1
      end
      yielded.should be_true
    end

    it 'should return block value' do
      @socket.should_receive(:send).with(match(/all@custom.metric.time:\d+/), 0)
      Metricsd::Client.record_time('custom.metric') do
        'hello'
      end.should eq('hello')
    end

    it 'should use options if time is not specified' do
      @socket.should_receive(:send).with(match(/all@custom.metric.time:\d+/), 0)
      yielded = false
      Metricsd::Client.record_time('custom.metric', {}) do
        yielded = true
      end
      yielded.should be_true
    end

    it 'should apply group if specified' do
      @socket.should_receive(:send).with('all@test.custom.metric.time:450', 0)
      Metricsd::Client.record_time('custom.metric', 0.45, :group => 'test')
    end

    it 'should apply source if empty string passed' do
      @socket.should_receive(:send).with('test@custom.metric.time:450', 0)
      Metricsd::Client.record_time('custom.metric', 0.45, :source => '')
    end

    it 'should apply source specified' do
      @socket.should_receive(:send).with('test2@custom.metric.time:450', 0)
      Metricsd::Client.record_time('custom.metric', 0.45, :source => 'test2')
    end
  end

  describe '.record_value' do
    it 'should record value' do
      @socket.should_receive(:send).with('all@custom.metric:23', 0)
      Metricsd::Client.record_value('custom.metric', 23)
    end

    it 'should apply group if specified' do
      @socket.should_receive(:send).with('all@test.custom.metric:23', 0)
      Metricsd::Client.record_value('custom.metric', 23, :group => 'test')
    end

    it 'should apply source if empty string passed' do
      @socket.should_receive(:send).with('test@custom.metric:23', 0)
      Metricsd::Client.record_value('custom.metric', 23, :source => '')
    end

    it 'should apply source specified' do
      @socket.should_receive(:send).with('test2@custom.metric:23', 0)
      Metricsd::Client.record_value('custom.metric', 23, :source => 'test2')
    end

    it 'should apply default group if specified' do
      Metricsd.default_group = 'grp'
      @socket.should_receive(:send).with('all@grp.custom.metric:23', 0)
      Metricsd::Client.record_value('custom.metric', 23)
    end

    it 'should override default group with the specified one' do
      Metricsd.default_group = 'grp'
      @socket.should_receive(:send).with('all@group.custom.metric:23', 0)
      Metricsd::Client.record_value('custom.metric', 23, :group => 'group')
    end

    it 'should clear group, if there is a default one, and empty string specified' do
      Metricsd.default_group = 'grp'
      @socket.should_receive(:send).with('all@custom.metric:23', 0)
      Metricsd::Client.record_value('custom.metric', 23, :group => '')
    end
  end

  describe '.record_values' do
    it 'should record all the values specified' do
      @socket.should_receive(:send).with('all@another.metric:47;all@custom.metric:23', 0)
      Metricsd::Client.record_values({'custom.metric' => 23, 'another.metric' => 47})
    end

    it 'should apply group if specified' do
      @socket.should_receive(:send).with('all@test.another.metric:47;all@test.custom.metric:23', 0)
      Metricsd::Client.record_values({'custom.metric' => 23, 'another.metric' => 47}, :group => 'test')
    end

    it 'should apply source if empty string passed' do
      @socket.should_receive(:send).with('test@another.metric:47;test@custom.metric:23', 0)
      Metricsd::Client.record_values({'custom.metric' => 23, 'another.metric' => 47}, :source => '')
    end

    it 'should apply source specified' do
      @socket.should_receive(:send).with('test2@another.metric:47;test2@custom.metric:23', 0)
      Metricsd::Client.record_values({'custom.metric' => 23, 'another.metric' => 47}, :source => 'test2')
    end

    it 'should split packets in 250 bytes' do
      packets, metrics = Array.new(2) { '' }, {}
      1.upto(8) do |metric|
        packet, name, value = (metric - 1) / 4, "my.super.long.custom.metric#{metric}", 10000000 + metric
        metrics[name] = value
        packets[packet] << "my.super.long.host.name@#{name}:#{value}" << (metric % 4 > 0 ? ';' : '')
      end

      @socket.should_receive(:send).once.with(packets[0], 0)
      @socket.should_receive(:send).once.with(packets[1], 0)
      Metricsd::Client.record_values(metrics, :source => 'my.super.long.host.name')
    end

    it 'should not allow packets larger than 250 bytes' do
      @socket.should_not_receive(:send)
      Metricsd.logger.should_receive(:warn).with(match(/Message is larger than 250 bytes/))
      Metricsd::Client.record_values({'a' * 300 => 5}, :source => 'my.super.long.host.name')
    end
  end
end


spec/metricsd_spec.rb
require 'bundler/setup'
require 'metricsd'

describe Metricsd do
  after :each do
    Metricsd.reset_defaults!
  end

  context 'with defaults' do
    source = Socket.gethostname.split('.').first

    it 'should have server_host = 127.0.0.1' do
      Metricsd.server_host.should == '127.0.0.1'
    end

    it 'should have server_port = 6311' do
      Metricsd.server_port.should == 6311
    end

    it "should have source = #{source}" do
      Metricsd.source.should == source
    end

    it 'should have default_source = all' do
      Metricsd.default_source.should == 'all'
    end

    it 'should have default_group = nil' do
      Metricsd.default_group.should be_nil
    end

    it 'should create logger' do
      Metricsd.logger.should be_a(Logger)
    end

    it 'should be enabled' do
      Metricsd.enabled?.should be_true
      Metricsd.enabled.should be_true
    end
  end

  context 'setters' do
    it 'should allow to change server_host' do
      Metricsd.server_host = 'metrics.local'
      Metricsd.server_host.should == 'metrics.local'
    end

    it 'should allow to change server_port' do
      Metricsd.server_port = '8000'
      Metricsd.server_port.should == 8000
    end

    it 'should not allow invalid server_port' do
      expect {
        Metricsd.server_port = 'aaa'
      }.to raise_error
    end

    it 'should allow to enabled and disable client' do
      Metricsd.enabled = false
      Metricsd.should_not be_enabled

      Metricsd.enable!
      Metricsd.should be_enabled

      Metricsd.disable!
      Metricsd.should_not be_enabled

      Metricsd.enabled = true
      Metricsd.should be_enabled
    end

    it 'should allow to change source' do
      Metricsd.source = 'test-host'
      Metricsd.source.should == 'test-host'
    end

    it 'should allow to change default_source' do
      Metricsd.default_source = 'nothing'
      Metricsd.default_source.should == 'nothing'
    end

    it 'should allow to change default_group' do
      Metricsd.default_group = 'mygroup'
      Metricsd.default_group.should == 'mygroup'
    end

    it 'should allow to change logger' do
      mock = Metricsd.logger = mock('Logger')
      Metricsd.logger.should be(mock)
    end
  end
end

lib/metricsd.rb
require 'socket'
require 'logger'
require 'benchmark'

# Pure Ruby client library for MetricsD server.
module Metricsd
  class << self
    # Gets the MetricsD server host. Default is "127.0.0.1".
    def server_host
      @@server_host
    end

    # Sets the MetricsD server host.
    def server_host=(host)
      @@server_host = host
      Client.reset_connection!
    end

    # Gets the MetricsD server port. Default is 6311.
    def server_port
      @@server_port
    end

    # Sets the MetricsD server port.
    def server_port=(port)
      @@server_port = Integer(port)
      Client.reset_connection!
    end

    # Get the value indicating whether Metricsd is enabled.
    def enabled?
      !!@@enabled
    end
    alias :enabled :enabled?

    # Sets the value indicating whether Metricsd is enabled.
    def enabled=(enabled)
      @@enabled = !!enabled
    end

    # Enables Metricsd client.
    def enable!
      @@enabled = true
    end

    # Disables Metricsd client.
    def disable!
      @@enabled = false
    end

    # Gets the source used to record host-specific metrics. Default is the
    # first part of the hostname (e.g. "test" for "test.host.com").
    def source
      @@source || metricsd.default_source
    end

    # Sets the source for host-specific metrics.
    def source=(source)
      @@source = source
    end

    # Gets the default source for all metrics. If nil or empty string — all
    # metrics will be host-specific (MetricsD server will generate per-host
    # graphs in addition to summary graph for all hosts for each metric).
    # Default is "all".
    def default_source
      @@default_source
    end

    # Sets the default source for all metrics.
    def default_source=(source)
      @@default_source = source
    end

    # Gets the default group for all metrics.
    def default_group
      @@default_group
    end

    # Sets the default group for all metrics.
    def default_group=(group)
      @@default_group = group
    end

    # Gets the logger used to output errors or warnings.
    def logger
      @@logger ||= if defined?(Rails)
        Rails.logger
      elsif defined?(Loops)
        Loops.logger
      else
        Logger.new(STDOUT)
      end
    end

    # Sets the logger used to output errors or warnings.
    def logger=(logger)
      @@logger = logger
    end

    # Resets all values to their default state (mostly for testing purpose).
    def reset_defaults!
      @@server_host    = '127.0.0.1'
      @@server_port    = 6311
      @@enabled        = true
      @@source         = Socket.gethostname[/^([^.]+)/, 1]
      @@default_source = 'all'
      @@default_group  = nil
      @@logger         = nil
    end
  end

  reset_defaults!
end

require 'metricsd/client'
require "metricsd/version"


lib/metricsd/client.rb
module Metricsd
  # Client class implements MetricsD protocol, allowing to send various metrics
  # to the MetricsD server for collecting, analyzing, and graphing them.
  #
  # Class allows to record count and timing stats to the database:
  #
  #   # Record success hit
  #   Metricsd::Client.record_success("api.docs.upload")
  #   # Record failure hit
  #   Metricsd::Client.record_failure("api.docs.upload")
  #   # Record timing info
  #   Metricsd::Client.record_time("api.docs.upload", 0.14)
  #   # Record complete success hit info (count + timing)
  #   Metricsd::Client.record_hit("api.docs.upload", true, 0.14)
  #   # Record an integer value
  #   Metricsd::Client.record_value("user.password.size", 15)
  #   Metricsd::Client.record_value("user.age", 26)
  #
  # To send several metrics in a single network packet, you can use record_values:
  #
  #   # Send all database pool stats
  #   Metricsd::Client.record_values({
  #     'db.pool.reserved'  => db_stats[:reserved],
  #     'db.pool.available' => db_stats[:available],
  #     'db.pool.pending'   => db_stats[:pending],
  #   }, :group => 'doc_timestamp')
  #
  # You can specify message source using <tt>:source => 'src'</tt> option. In this case
  # you will be able to see summary graphs and graphs per source:
  #
  #   # Generate graphs for all tables, and each single table.
  #   Metricsd::Client.record_success("hbase.reads", :source => @hbase_table)
  #
  # By default only summary statistics is calculated. You can enable per-host graphs
  # by specifying the appropriate source:
  #
  #   # Generate summary graph for all hosts, and graphs for each single host.
  #   Metricsd::Client.record_success("hbase.reads", :source => Metricsd::Client.source)
  #   # ... or you can pass an empty string with the same effect.
  #   Metricsd::Client.record_success("hbase.reads", :source => '')
  #
  # You can group your metrics using <tt>:group</tt> option. In this case metrics will be
  # displayed together on the summary page.
  #
  #   # Group metrics using :group option.
  #   Metricsd::Client.record_success("reads", :source => @hbase_table, :group => 'hbase')
  #   # Group metrics using special syntax "group.metric".
  #   Metricsd::Client.record_success("hbase.reads", :source => @hbase_table)
  #
  class Client
    class << self
      # Record complete hit info. Time should be a floating point
      # number of seconds.
      #
      # It creates two metrics:
      # * +your.metric.status+ with counts of failed and succeded events
      # * +your.metric.time+ with time statistics
      #
      # @param [String] metric is the metric name (like app.docs.upload)
      # @param [Boolean] is_success indicating whether request was successful.
      # @param [Float] time floating point number of seconds.
      # @param [Hash] opts options.
      # @option opts [String] :group metrics group.
      # @option opts [String] :source metric source.
      #
      def record_hit(metric, is_success, time, opts = {})
        record_internal({
            "#{metric}.status" => is_success ? 1 : -1,
            "#{metric}.time"   => (time * 1000).round
          }, opts
        )
      end
      alias :hit :record_hit

      # Record succeded boolean event.
      #
      # It creates a single metric:
      # * +your.metric.status+ with numbers of failed and succeded events
      #
      # @param [String] metric is the metric name (like app.docs.upload)
      # @param [Hash] opts options.
      # @option opts [String] :group metrics group.
      # @option opts [String] :source metric source.
      #
      def record_success(metric, opts = {})
        record_status(metric, true, opts)
      end
      alias :success :record_success

      # Record failed boolean event.
      #
      # It creates a single metric:
      # * +your.metric.status+ with numbers of failed and succeded events
      #
      # @param [String] metric is the metric name (like app.docs.upload)
      # @param [Hash] opts options.
      # @option opts [String] :group metrics group.
      # @option opts [String] :source metric source.
      #
      def record_failure(metric, opts = {})
        record_status(metric, false, opts)
      end
      alias :failure :record_failure

      # Record status event (success or failure).
      #
      # It creates a single metric:
      # * +your.metric.status+ with numbers of failed and succeded events
      #
      # @param [String] metric is the metric name (like app.docs.upload)
      # @param [Boolean] is_success indicating whether event is successful.
      # @param [Hash] opts options.
      # @option opts [String] :group metrics group.
      # @option opts [String] :source metric source.
      #
      def record_status(metric, is_success, opts = {})
        record_internal({"#{metric}.status" => is_success ? 1 : -1}, opts)
      end
      alias :status :record_status

      # Record timing info. Time should be a floating point
      # number of seconds.
      #
      # It creates a single metric:
      # * +your.metric.time+ with time statistics
      #
      # @param [String] metric is the metric name (like app.docs.upload)
      # @param [Float] time floating point number of seconds.
      # @param [Hash] opts options.
      # @option opts [String] :group metrics group.
      # @option opts [String] :source metric source.
      #
      def record_time(metric, time = nil, opts = {}, &block)
        opts, time = time, nil if Hash === time
        result = nil
        if time.nil?
          raise ArgumentError, "You should pass a block if time is not given" unless block_given?
          time = Benchmark.measure do
            result = block.call
          end.real
        end
        record_internal({"#{metric}.time" => (time * 1000).round}, opts)
        result
      end
      alias :time :record_time

      # Record an integer value.
      #
      # It creates a single metric:
      # * +your.metric+ with values statistics
      #
      # @param [String] metric is the metric name (like app.docs.upload)
      # @param [Integer] value metric value.
      # @param [Hash] opts options.
      # @option opts [String] :group metrics group.
      # @option opts [String] :source metric source.
      #
      def record_value(metric, value, opts = {})
        record_internal({metric => value.round}, opts)
      end
      alias :record :record_value
      alias :value :record_value

      # Record multiple integer values.
      #
      # It creates a metric for each entry in +metrics+ Hash:
      # * +your.metric+ with values statistics
      #
      # @param [Hash] metrics a +Hash+ that maps metrics names to their values.
      # @param [Hash] opts options.
      # @option opts [String] :group metrics group.
      # @option opts [String] :source metric source.
      #
      # @example
      #   Metricsd::Client.record_values(
      #     'db.pool.reserved'  => db_stats[:reserved],
      #     'db.pool.available' => db_stats[:available],
      #     'db.pool.pending'   => db_stats[:pending],
      #   )
      #
      def record_values(metrics, opts = {})
        record_internal(metrics, opts)
      end
      alias :values :record_values

      # Reset and re-establish connection.
      def reset_connection!
        @@socket = nil
      end

    private

      # Returns a UDP socket used to send metrics to MetricsD.
      def collector_socket
        @@lock.synchronize do
          @@socket ||= UDPSocket.new.tap do |sock|
            sock.connect(Metricsd.server_host, Metricsd.server_port)
          end
        end
      rescue SocketError => e
        Metricsd.logger.error("Exception occurred while trying to connect to MetricsD (#{Metricsd.server_host}:#{Metricsd.server_port}): #{e.inspect}")
        nil
      end

      # Send informating to the RRD collector daemon using UDP protocol.
      def record_internal(metrics, opts = {})
        return unless Metricsd.enabled?

        opts = { :source => Metricsd.default_source }.update(opts)
        opts[:source] = Metricsd.source if opts[:source].empty?

        # Build the message for
        send_in_packets Array(metrics).map { |arg| pack(arg[0], arg[1], opts) }.sort
      end

      # Combines string representations of metrics into packets of 250 bytes and
      # sends them to MetricsD.
      def send_in_packets(strings)
        msg = ''
        strings.each do |s|
          if s.size > 250
            Metricsd.logger.warn("Message is larger than 250 bytes, so it was ignored: #{s}")
            next
          end

          if msg.size + s.size + (msg.size > 0 ? 1 : 0) > 250
            safe_send(msg)
            msg = ''
          end
          msg << (msg.size > 0 ? ';' : '') << s
        end
        safe_send(msg) if msg.size > 0
      end

      # Sends a string to the MetricsD. Should never raise any network-specific
      # exceptions, but log them instead, and silently return.
      def safe_send(msg)
        # Check if we could connect to the MetricsD
        return false unless collector_socket
        collector_socket.send(msg, 0)
        true
      rescue Errno::ECONNREFUSED => e
        Metricsd.logger.error("Exception occurred while trying to send data to MetricsD (#{Metricsd.server_host}:#{Metricsd.server_port}): #{e.inspect}")
        false
      end

      # Packs metric into a string representation according to the MetricsD
      # protocol.
      def pack(key, value, opts)
        group = opts[:group] || Metricsd.default_group || ''
        key = "#{group}.#{key}" unless group.empty?
        opts[:source].empty? ? "#{key}:#{value}" : "#{opts[:source]}@#{key}:#{value}"
      end
    end
    @@lock = Monitor.new
  end
end
